### PR TITLE
Nullable annotate ListViewGroup

### PR DESCRIPTION
**Pull request by hughbe**
## Proposed Changes
- Nullable annotate ListViewGroup
- Fix NRE bug

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/2828)

### CODE REVIEW

**Comment by RussKie** [comment]
Just a heads up: we're currently blocked by our downstream consumers and unable to update to the latest .NET Runtime, which contains the latest compiler needed for the Roslyn analysers.
It maybe another month or so... I'll provide an update when the situation changes.

**Comment by RussKie** [comment]
You need to apply PubllicAPI analyzer fixes to record the changes to the public API surface.

**Comment by hughbe** [comment]
Right... So I *think* I've figured out how to do this.

For my sanity:
- "~" before each line does something: what it does, god knows. I have a hunch it disables nullable annotations? I cannot find anything annotated for my sanity
- You need to explicitly annotate non-null parameters with `!` at the end

There's one more things that I think is a bug in the analyzer. There is an error
```
System/Windows/Forms/ListViewGroup.cs(168,16): warning RS0036: Symbol 'Items.get' is missing nullability annotations in the declared API. [/Users/hugh/Documents/GitHub/winforms/src/System.Windows.Forms/src/System.Windows.Forms.csproj]
```

`ListViewGroup.Items` is defined as non-nullable in c#
```cs
public ListView.ListViewItemCollection Items
    => _items ??= new ListView.ListViewItemCollection(new ListViewGroupItemCollection(this));
```

and in the `PublicApi.Shipped.txt":
```
System.Windows.Forms.ListViewGroup.Items.get -> System.Windows.Forms.ListView.ListViewItemCollection!
```

So I'm confused why this error occurs

**Please** can there be some sort of option to just regenerate `PublicApi.Shipped.txt`. To be honest, there might be but there doesn't seem to be any docs :))

**Comment by hughbe** [comment]
Yup. This is very obviously a bug. See the latest commit for the fix. I am guessing the analyzer doesn't understand `??=` syntax

**Comment by jcouv** [comment]
@hughbe @RussKie 

> "~" before each line does something

This marker just documents that this API still mentions an oblivious type. The purpose is to help track annotation effort (ie. remove all those markers).
I'm adding [some documentation](https://github.com/dotnet/roslyn-analyzers/blob/master/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md#nullable-reference-type-support) as I see the kinds of questions you have. Let me know if still needs improvement. Thanks for your patience :-)

> You need to explicitly annotate non-null parameters with ! at the end

Yes, when the publicAPI file includes `#nullable enable`, then `!` means non-nullable. This contrasts with "" (blank) which would mean oblivious. And `?` means nullable.

> error related to  `System.Windows.Forms.ListViewGroup.Items.get`

I'm confused too. Just to confirm, was the line in publicAPI file (`System.Windows.Forms.ListViewGroup.Items.get -> System.Windows.Forms.ListView.ListViewItemCollection!`) generated by publicAPI analyzer/fixer, or did you edit it by hand (I would recommend against that in general).

Let's sync on Monday. We can do LiveShare or Teams meeting to troubleshoot.

> See the latest commit for the fix. I am guessing the analyzer doesn't understand ??= syntax

If you mean the publicAPI analyzer, then I don't think so. That analyzer only looks at public APIs, not the bodies/implementations of methods.

**Comment by jcouv** [comment]
> Please can there be some sort of option to just regenerate PublicApi.Shipped.txt. 

Depending on the warning, a fix should be offered at the location of the warning. Select the desired fix (with Project or Solution scope) and that should be it.

File: src/System.Windows.Forms/src/System/Windows/Forms/ListViewGroup.cs [hunk 1]
```diff
@@ -49,7 +47,7 @@ private ListViewGroup(SerializationInfo info, StreamingContext context) : this()
         /// <summary>
         ///  Creates a ListViewItem object from a Key and a Name
         /// </summary>
-        public ListViewGroup(string key, string headerText) : this()
+        public ListViewGroup(string? key, string headerText) : this()
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListViewGroup.cs:50]: Should it be this?
```suggestion
        public ListViewGroup(string? key, string? headerText) : this()
```

File: src/System.Windows.Forms/src/PublicAPI.Shipped.txt [hunk 2]
```diff
@@ -4365,24 +4365,24 @@ System.Windows.Forms.ListViewAlignment.Left = 1 -> System.Windows.Forms.ListView
...
-~System.Windows.Forms.ListViewGroup.ListViewGroup(string header) -> void
-~System.Windows.Forms.ListViewGroup.ListViewGroup(string header, System.Windows.Forms.HorizontalAlignment headerAlignment) -> void
-~System.Windows.Forms.ListViewGroup.ListViewGroup(string key, string headerText) -> void
-~System.Windows.Forms.ListViewGroup.Name.get -> string
-~System.Windows.Forms.ListViewGroup.Name.set -> void
-~System.Windows.Forms.ListViewGroup.Tag.get -> object
-~System.Windows.Forms.ListViewGroup.Tag.set -> void
+System.Windows.Forms.ListViewGroup.ListViewGroup(string! header) -> void
+System.Windows.Forms.ListViewGroup.ListViewGroup(string! header, System.Windows.Forms.HorizontalAlignment headerAlignment) -> void
+System.Windows.Forms.ListViewGroup.ListViewGroup(string? key, string! headerText) -> void
```
Comment (RussKie) [inline src/System.Windows.Forms/src/PublicAPI.Shipped.txt:4381]: I think these should be:
```suggestion
System.Windows.Forms.ListViewGroup.ListViewGroup(string? header) -> void
System.Windows.Forms.ListViewGroup.ListViewGroup(string? header, System.Windows.Forms.HorizontalAlignment headerAlignment) -> void
System.Windows.Forms.ListViewGroup.ListViewGroup(string? key, string? headerText) -> void
```

**Comment by RussKie** [comment]
> Let's sync on Monday. We can do LiveShare or Teams meeting to troubleshoot.

I don't think it is possible, PST, GMT and AEST don't overlap... I'd like to sync up about https://github.com/dotnet/roslyn-analyzers/issues/3582

**Comment by hughbe** [comment]
Yeah you're right. I've updated ListViewGroup.Header/Footer to have `AllowNull` attributes (meaning that you can set null but it never returns null`) and made the header parameters in the constructors nullable

**Comment by RussKie** [comment]
The builds is failing complaining about annotations.... Right now NRT annotation feel like a tremendous up hill battle. 😩

**Comment by hughbe** [comment]
I know... I may be blind, but I don't get the error
> System\Windows\Forms\ListViewGroup.cs#L170
System\Windows\Forms\ListViewGroup.cs(170,57): error RS0036: (NETCORE_ENGINEERING_TELEMETRY=Build) Symbol 'Items.get' is missing nullability annotations in the declared API.

**Comment by weltkante** [comment]
I played around with this a bit, looks like the tool cannot handle nested classes as return values. Probably needs to be raised as an issue somewhere else.

**Comment by hughbe** [comment]
I suspected as much!

**Comment by jcouv** [comment]
> The builds is failing complaining about annotations.... Right now NRT annotation feel like a tremendous up hill battle. 

Consider annotating your code without putting `#nullable enable` in the public API file for now.
I'll look into the nested type issue. I thought that was [fixed already](https://github.com/dotnet/roslyn-analyzers/pull/3425).

**Comment by jcouv** [comment]
Confirmed the bug with return of nested type. Will fix asap

**Comment by jcouv** [comment]
Hum, seems like I spoke too soon. I cannot repro the issue (found a bug with a compiler API, but that's not the one used by the publicAPI analyzer...). Below is the test to verify that.

I wonder if you may be using a version that predates [this fix](https://github.com/dotnet/roslyn-analyzers/pull/3425). I'll take a look at your package versions and get back to you. 

``` csharp
        [Fact]
        public async Task AnnotatedMemberInAnnotatedShippedAPI()
        {
            var source = @"
#nullable enable
public class C
{
    public string? OldField;
    public string? {|RS0036:Field|};
    public string {|RS0036:Field2|};
    public Nested? {|RS0036:Field3|};
    public Nested {|RS0036:Field4|};
    public class Nested { }
}
";

            var shippedText = @"#nullable enable
C
C.C() -> void
C.Nested
C.Nested.Nested() -> void
C.OldField -> string?
C.Field -> string
C.Field2 -> string
C.Field3 -> C.Nested
C.Field4 -> C.Nested";

            var unshippedText = @"";

            var fixedShippedText = @"#nullable enable
C
C.C() -> void
C.Nested
C.Nested.Nested() -> void
C.OldField -> string?
C.Field -> string?
C.Field2 -> string!
C.Field3 -> C.Nested?
C.Field4 -> C.Nested!";

            await VerifyCSharpAdditionalFileFixAsync(source, shippedText, unshippedText, fixedShippedText, newUnshippedApiText: unshippedText);
        }
```

**Comment by jcouv** [comment]
Alright, I dug into this. It is indeed a problem that was fixed in more recent publicAPI analyzer. I didn't trace the fix down.
Here's a change that I made locally and built successfully: https://github.com/dotnet/winforms/commit/049f256bd47da9b17bc35ac762be1588b7fdd2e8

@mavasani I noticed that the most recent publicAPI analyzer on nuget is [3.0](https://www.nuget.org/packages/Microsoft.CodeAnalysis.PublicApiAnalyzers/3.0.0). But on myget, we're up to [3.3-beta](https://dotnet.myget.org/feed/roslyn-analyzers/package/nuget/Microsoft.CodeAnalysis.PublicApiAnalyzers).
Could we publish a 3.2 or a 3.3-beta so that winforms repo can reference it without adding myget as a source?

**Comment by mavasani** [comment]
Just published https://www.nuget.org/packages/Microsoft.CodeAnalysis.PublicApiAnalyzers/3.3.0-beta1.final, which should have all the changes to this analyzer.

**Comment by jcouv** [comment]
Thanks @mavasani :-)

@hughbe I've pushed a commit that updates the roslyn-analyzer package reference and fixes publicAPI file. I hope that's okay.

**Comment by RussKie** [comment]
Looks like the build agent is unable to find the analyzer:
```
##[error]src\System.Windows.Forms\src\System.Windows.Forms.csproj(0,0): error NU1603: (NETCORE_ENGINEERING_TELEMETRY=Restore) 
System.Windows.Forms depends on Microsoft.CodeAnalysis.PublicApiAnalyzers (>= 3.3.0-beta1.20255.4) but Microsoft.CodeAnalysis.PublicApiAnalyzers 3.3.0-beta1.20255.4 was not found. An approximate best match of Microsoft.CodeAnalysis.PublicApiAnalyzers 3.3.0-beta1.final was resolved.
```

**Comment by jcouv** [comment]
I'd accidentally reverted the version with a wrong `checkout` operation... Pushed a commit to fix that. Thanks

**Comment by RussKie** [comment]
Thank you!

**Comment by RussKie** [comment]
/cc: @Logerfo looks like we are starting to make progress

**Comment by RussKie** [comment]
Docs: https://github.com/dotnet/docs/issues/17085

**Pull request merged**